File: Pokemon.cpp
#include "Pokemon.h"
#include <iostream>
#include "global.h"

// Implementación de métodos
Pokemon::Pokemon(std::string n, std::vector<Tipo> t, int v, int a, int d)
    : nombre(n), tipos(t), vida(v), vidaMax(v), ataque(a), defensa(d) {}

std::string Pokemon::getNombre() const { return nombre; }
int Pokemon::getVida() const { return vida; }
int Pokemon::getVidaMax() const { return vidaMax; }
int Pokemon::getAtaque() const { return ataque; }
int Pokemon::getDefensa() const { return defensa; }
bool Pokemon::estaVivo() const { return vida > 0; }

void Pokemon::agregarMovimiento(const Movimiento& m) { movimientos.push_back(m); }
Movimiento Pokemon::getMovimiento(int index) const { return movimientos.at(index); }

const std::vector<Tipo>& Pokemon::getTipos() const { return tipos; }

const std::vector<Movimiento>& Pokemon::getMovimientos() const {
    return movimientos;
}

void Pokemon::mostrar() const {
    std::cout << nombre << " [" << vida << "/" << vidaMax << "]";
}

void Pokemon::recibirDaño(int dmg) {
    vida -= dmg;
    if (vida < 0) vida = 0;
}

void Pokemon::curar(int cantidad) {
    vida += cantidad;
    if (vida > vidaMax) vida = vidaMax;
}

// void Pokemon::atacar(Pokemon& objetivo) {
//     int dmg = ataque - (objetivo.defensa / 2);
//     if (dmg < 1) dmg = 1;
//     cout << nombre << " ataca causando " << dmg << " de daño a " 
//          << objetivo.nombre << "." << endl;
//     objetivo.recibirDaño(dmg);
// }

void Pokemon::atacar(Pokemon& objetivo, const Movimiento& movimiento) {
    double efectividadTotal = 1.0;
    
    // Multiplica la efectividad para cada uno de los tipos
    for (Tipo tipoDefensor : objetivo.getTipos()) {
        efectividadTotal *= calcularEfectividad(movimiento.getTipo(), tipoDefensor);
    }
    
    int dmg = (ataque + movimiento.getPotencia()) - (objetivo.getDefensa() / 2);
    if (dmg < 1) dmg = 1;

    dmg *= efectividadTotal;

    std::cout << nombre << " usa " << movimiento.getNombre() << " y causa " << dmg << " de danio a " 
              << objetivo.getNombre() << "." << std::endl;
    
    // Mensajes de efectividad
    if (efectividadTotal >= 2.0) {
        std::cout << YELLOW << "Es super efectivo!" << RESET << std::endl;
    } else if (efectividadTotal < 1.0 && efectividadTotal > 0.0) {
        std::cout << CYAN << "No es muy efectivo..." << RESET << std::endl;
    } else if (efectividadTotal == 0.0) {
        std::cout << CYAN << "No tiene efecto..." << RESET << std::endl;
    }

    objetivo.recibirDaño(dmg);
}

File: pokemon/Pokemon.java
package pokemon;

import pokemon.movimientos.Movimiento;
import pokemon.tipos.TipoElemento;
import java.util.ArrayList;
import java.util.List;

//Clase que representa un Pokemon en el sistema de combate

public class Pokemon {
    private String nombre;
    private List<TipoElemento> tipos;
    private List<Movimiento> movimientos;
    private int vida;
    private int vidaMax;
    private int ataque;
    private int defensa;
    private String spritePath; //para su imagen
    
    public Pokemon(String nombre, List<TipoElemento> tipos, int vida, int ataque, int defensa) {
        this.nombre = nombre;
        this.tipos = new ArrayList<>(tipos);
        this.movimientos = new ArrayList<>();
        this.vida = vida;
        this.vidaMax = vida;
        this.ataque = ataque;
        this.defensa = defensa;
        
    }
    
    public void agregarMovimiento(Movimiento movimiento) {
        movimientos.add(movimiento);
    }
    
    public Movimiento getMovimiento(int index) {
        return movimientos.get(index);
    }
    
    // Getters
    public String getNombre() {
        return nombre;
    }
    
    public int getVida() {
        return vida;
    }
    
    public int getVidaMax() {
        return vidaMax;
    }
    
    public int getAtaque() {
        return ataque;
    }
    
    public int getDefensa() {
        return defensa;
    }
    public String getSpritePath() {
        return spritePath;
    }
    public void setSpritePath(String spritePath) {
        this.spritePath = spritePath;
    }

    public boolean estaVivo() {
        return vida > 0;
    }
    
    public List<TipoElemento> getTipos() {
        return new ArrayList<>(tipos); // Retorna copia defensiva
    }
    
    public List<Movimiento> getMovimientos() {
        return new ArrayList<>(movimientos); // Retorna copia defensiva
    }
    
    public void mostrar() {
        System.out.print(nombre + " [" + vida + "/" + vidaMax + "]");
    }
    
    public void recibirDano(int dmg) {
        vida -= dmg;
        if (vida < 0) vida = 0;
    }
    
    public void curar(int cantidad) {
        vida += cantidad;
        if (vida > vidaMax) vida = vidaMax;
    }
    
    public void atacar(Pokemon objetivo, Movimiento movimiento) {
        movimiento.aplicarEfecto(this, objetivo);
    }
}